File: surface/widgets/compose/components/widgets/feedback/Tooltip.kt
// ============================================================
// Clef Surface Compose Widget — Tooltip
//
// Lightweight floating label that provides supplementary
// descriptive text for a trigger element. Shown on long-press
// or pointer hover. Tooltips are non-interactive -- they
// contain no focusable elements.
//
// Compose adaptation: Material 3 PlainTooltip wrapped in a
// TooltipBox with a TooltipState. Supports top and bottom
// placement via TooltipDefaults.
// See widget spec: repertoire/widgets/feedback/tooltip.widget
// ============================================================

package clef.surface.compose.components.widgets.feedback

import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier

// --------------- Component ---------------

/**
 * Lightweight floating label providing supplementary descriptive text.
 *
 * @param content Descriptive text displayed in the tooltip.
 * @param modifier Modifier applied to the TooltipBox.
 * @param trigger Trigger element that anchors the tooltip.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ClefTooltip(
    content: String,
    modifier: Modifier = Modifier,
    trigger: @Composable () -> Unit = {},
) {
    val tooltipState = rememberTooltipState()

    TooltipBox(
        positionProvider = TooltipDefaults.rememberPlainTooltipPositionProvider(),
        tooltip = {
            PlainTooltip {
                Text(text = content)
            }
        },
        state = tooltipState,
        modifier = modifier,
    ) {
        trigger()
    }
}

/**
 * Rich tooltip variant with title and descriptive body text.
 *
 * @param title Heading text displayed prominently in the tooltip.
 * @param description Body text with additional details.
 * @param actionLabel Optional action label rendered as a clickable text button.
 * @param onAction Callback fired when the action is activated.
 * @param modifier Modifier applied to the TooltipBox.
 * @param trigger Trigger element that anchors the tooltip.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ClefRichTooltip(
    title: String,
    description: String,
    actionLabel: String? = null,
    onAction: (() -> Unit)? = null,
    modifier: Modifier = Modifier,
    trigger: @Composable () -> Unit = {},
) {
    val tooltipState = rememberTooltipState(isPersistent = actionLabel != null)

    TooltipBox(
        positionProvider = TooltipDefaults.rememberRichTooltipPositionProvider(),
        tooltip = {
            RichTooltip(
                title = { Text(text = title) },
                action = if (actionLabel != null && onAction != null) {
                    {
                        TextButton(onClick = onAction) {
                            Text(text = actionLabel)
                        }
                    }
                } else {
                    null
                },
            ) {
                Text(text = description)
            }
        },
        state = tooltipState,
        modifier = modifier,
    ) {
        trigger()
    }
}

File: surface/widgets/swiftui/components/widgets/feedback/Tooltip.swift
// ============================================================
// Clef Surface SwiftUI Widget — Tooltip
//
// Lightweight floating label that provides supplementary
// descriptive text for a trigger element. In SwiftUI, uses
// the .help modifier for basic tooltips and a popover for
// rich tooltips.
//
// Adapts the tooltip.widget spec to SwiftUI rendering.
// ============================================================

import SwiftUI

// --------------- Component ---------------

/// Tooltip view providing supplementary descriptive text.
///
/// - Parameters:
///   - content: Descriptive text displayed in the tooltip.
///   - trigger: Trigger element that anchors the tooltip.
struct TooltipView<Trigger: View>: View {
    var content: String
    @ViewBuilder var trigger: Trigger

    var body: some View {
        trigger
            .help(content)
    }
}

/// Rich tooltip variant with title and descriptive body text.
///
/// - Parameters:
///   - title: Heading text displayed in the tooltip.
///   - description: Body text with additional details.
///   - actionLabel: Optional action label text.
///   - onAction: Callback fired when the action is activated.
///   - trigger: Trigger element that anchors the tooltip.
struct RichTooltipView<Trigger: View>: View {
    var title: String
    var description: String
    var actionLabel: String? = nil
    var onAction: (() -> Void)? = nil
    @ViewBuilder var trigger: Trigger

    @State private var isShowing = false

    var body: some View {
        trigger
            .onLongPressGesture {
                isShowing = true
            }
            .popover(isPresented: $isShowing) {
                VStack(alignment: .leading, spacing: 8) {
                    Text(title)
                        .font(.subheadline)
                        .fontWeight(.semibold)
                    Text(description)
                        .font(.caption)
                        .foregroundColor(.secondary)
                    if let actionLabel = actionLabel, let onAction = onAction {
                        SwiftUI.Button(actionLabel, action: onAction)
                            .font(.caption)
                    }
                }
                .padding(12)
            }
    }
}
